x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java

```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
             }
         }
 
-        advance(currentStage, listener);
+        int c = currentStage;
+        checkMissingEvents(() -> advance(c, listener), listener);
     }
 
     /**
      * Move the window while advancing the query base.
      */
     private void rebaseWindow(int nextStage, ActionListener<Payload> listener) {
         log.trace("Rebasing window...");
-        advance(nextStage, listener);
+        checkMissingEvents(() -> advance(nextStage, listener), listener);
+    }
+
+    public void checkMissingEvents(Runnable next, ActionListener<Payload> listener) {
+        Set<Sequence> sequencesToCheck = matcher.toCheckForMissing();
+        if (sequencesToCheck.isEmpty()) {
+            if (matcher.limitReached()) {
+                doPayload(listener);
+                return;
+            }
+            next.run();
+        } else {
+            Iterator<Sequence> iterator = sequencesToCheck.iterator();
+            List<Sequence> batchToCheck = new ArrayList<>();
+
+            for (int i = 0; i < MAX_SEQUENCES_TO_CHECK_FOR_MISSING && iterator.hasNext(); i++) {
+                batchToCheck.add(iterator.next());
+                iterator.remove();
+            }
+
+            List<SearchRequest> queries = prepareQueryForMissingEvents(batchToCheck);
+            client.multiQuery(queries, wrap(p -> doCheckMissingEvents(batchToCheck, p, listener, next), listener::onFailure));
+        }
+    }
+
+    public void doCheckMissingEvents(List<Sequence> batchToCheck, MultiSearchResponse p, ActionListener<Payload> listener, Runnable next) {
+        MultiSearchResponse.Item[] responses = p.getResponses();
+        int nextResponse = 0;
+        for (Sequence sequence : batchToCheck) {
+            boolean leading = true;
+            boolean discarded = false;
+            Timestamp lastLeading = null;
+            Timestamp firstTrailing = null;
+            for (int i = 0; i < criteria.size(); i++) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    SearchResponse response = responses[nextResponse++].getResponse();
+                    if (discarded) {
+                        continue; // consume all the responses for this sequence even if it's already discarded
+                    }
+                    SearchHit[] hits = response.getHits().getHits();
+                    if (leading) {
+                        if (hits.length == 0) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        lastLeading = lastLeading == null || lastLeading.delta(hitTimestamp) < 0 ? hitTimestamp : lastLeading;
```

Comment:
Why the choice of `delta` method here? You don't actually need the delta (which is calculated differently) but if one is earlier than the other. Which means a `compareTo` is enough: `lastLeading.instant().compareTo(hitTimestamp.instant())`. There are slight differences between two approaches, but if I'm not missing something the compareTo above is still using the nanos precision.

Reply:
This one is not reflected in the changes.